File: app/src/main/java/com/simplecity/amp_library/search/SearchPresenter.java
package com.simplecity.amp_library.search;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import com.annimon.stream.Stream;
import com.simplecity.amp_library.model.Album;
import com.simplecity.amp_library.model.AlbumArtist;
import com.simplecity.amp_library.model.Song;
import com.simplecity.amp_library.playback.MediaManager;
import com.simplecity.amp_library.ui.modelviews.AlbumArtistView;
import com.simplecity.amp_library.ui.modelviews.AlbumView;
import com.simplecity.amp_library.ui.presenters.Presenter;
import com.simplecity.amp_library.utils.DataManager;
import com.simplecity.amp_library.utils.LogUtils;
import com.simplecity.amp_library.utils.SettingsManager;
import com.simplecity.amp_library.utils.StringUtils;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.SingleOperator;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;
import kotlin.Unit;

public class SearchPresenter extends Presenter<SearchView> {

    private static final String TAG = "SearchPresenter";

    private static final double SCORE_THRESHOLD = 0.80;

    private Disposable performSearchSubscription;

    private String query;
    private MediaManager mediaManager;

    @Inject
    public SearchPresenter(@NonNull MediaManager mediaManager) {
        this.mediaManager = mediaManager;
    }

    @Override
    public void bindView(@NonNull SearchView view) {
        super.bindView(view);

        view.setFilterFuzzyChecked(SettingsManager.getInstance().getSearchFuzzy());
        view.setFilterArtistsChecked(SettingsManager.getInstance().getSearchArtists());
        view.setFilterAlbumsChecked(SettingsManager.getInstance().getSearchAlbums());
    }

    @Override
    public void unbindView(@NonNull SearchView view) {
        super.unbindView(view);
    }

    void queryChanged(@Nullable String query) {

        if (TextUtils.isEmpty(query)) {
            query = "";
        }

        if (query.equals(this.query)) {
            return;
        }

        loadData(query);

        this.query = query;
    }

    private void loadData(@NonNull String query) {

        SearchView searchView = getView();

        if (searchView != null) {

            searchView.setLoading(true);

            //We've received a new refresh call. Unsubscribe the in-flight subscription if it exists.
            if (performSearchSubscription != null) {
                performSearchSubscription.dispose();
            }

            Single<List<AlbumArtist>> albumArtistsObservable = SettingsManager.getInstance().getSearchArtists() ? DataManager.getInstance().getAlbumArtistsRelay()
                    .first(Collections.emptyList())
                    .lift(new AlbumArtistFilterOperator(query)) : Single.just(Collections.emptyList());

            Single<List<Album>> albumsObservable = SettingsManager.getInstance().getSearchAlbums() ? DataManager.getInstance().getAlbumsRelay()
                    .first(Collections.emptyList())
                    .lift(new AlbumFilterOperator(query)) : Single.just(Collections.emptyList());

            Single<List<Song>> songsObservable = DataManager.getInstance().getSongsRelay()
                    .first(Collections.emptyList())
                    .lift(new SongFilterOperator(query));

            performSearchSubscription = Single.zip(albumArtistsObservable, albumsObservable, songsObservable, SearchResult::new)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            searchView::setData,
                            error -> LogUtils.logException(TAG, "Error refreshing adapter", error)
                    );

            addDisposable(performSearchSubscription);
        }
    }

    void setSearchFuzzy(boolean searchFuzzy) {
        SettingsManager.getInstance().setSearchFuzzy(searchFuzzy);
        loadData(query);
    }

    void setSearchArtists(boolean searchArtists) {
        SettingsManager.getInstance().setSearchArtists(searchArtists);
        loadData(query);
    }

    void setSearchAlbums(boolean searchAlbums) {
        SettingsManager.getInstance().setSearchAlbums(searchAlbums);
        loadData(query);
    }

    void onSongClick(List<Song> songs, Song song) {
        SearchView view = getView();

        mediaManager.playAll(songs, songs.indexOf(song), true, message -> {
            if (view != null) {
                view.showToast(message);
            }
            return Unit.INSTANCE;
        });
    }

    void onArtistClicked(AlbumArtistView albumArtistView, AlbumArtistView.ViewHolder viewholder) {
        com.simplecity.amp_library.search.SearchView view = getView();
        if (view != null) {
            view.goToArtist(albumArtistView.albumArtist, viewholder.imageOne);
        }
    }

    void onAlbumClick(AlbumView albumView, AlbumView.ViewHolder viewHolder) {
        com.simplecity.amp_library.search.SearchView view = getView();
        if (view != null) {
            view.goToAlbum(albumView.album, viewHolder.imageOne);
        }
    }

    private class SongFilterOperator implements SingleOperator<List<Song>, List<Song>> {

        String filterString;

        SongFilterOperator(@NonNull String filterString) {
            this.filterString = filterString;
        }

        @Override
        public SingleObserver<? super List<Song>> apply(SingleObserver<? super List<Song>> observer) {
            return new SingleObserver<List<Song>>() {
                @Override
                public void onSubscribe(Disposable d) {
                    observer.onSubscribe(d);
                }

                @Override
                public void onSuccess(List<Song> songs) {
                    Stream<Song> songStream = Stream.of(songs).filter(song -> song.name != null);
                    songs = (SettingsManager.getInstance().getSearchFuzzy() ? applyJaroWinklerFilter(songStream) : applySongFilter(songStream)).toList();
                    observer.onSuccess(songs);
                }

                @Override
                public void onError(Throwable e) {
                    observer.onError(e);
                }
            };
        }

        Stream<Song> applyJaroWinklerFilter(Stream<Song> songStream) {
            return songStream.map(song -> new SearchUtils.JaroWinklerObject<>(song, filterString, song.name))
                    .filter(jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString))
                    .sorted((a, b) -> a.object.compareTo(b.object))
                    .sorted((a, b) -> Double.compare(b.score, a.score))
                    .map(jaroWinklerObject -> jaroWinklerObject.object);
        }

        Stream<Song> applySongFilter(Stream<Song> songStream) {
            return songStream.filter(song -> StringUtils.containsIgnoreCase(song.name, filterString));
        }
    }

    private class AlbumFilterOperator implements SingleOperator<List<Album>, List<Album>> {

        String filterString;

        AlbumFilterOperator(@NonNull String filterString) {
            this.filterString = filterString;
        }

        @Override
        public SingleObserver<? super List<Album>> apply(SingleObserver<? super List<Album>> observer) {
            return new SingleObserver<List<Album>>() {
                @Override
                public void onSubscribe(Disposable d) {
                    observer.onSubscribe(d);
                }

                @Override
                public void onSuccess(List<Album> albums) {
                    Collections.sort(albums, Album::compareTo);
                    Stream<Album> albumStream = Stream.of(albums).filter(album -> album.name != null);
                    Stream<Album> filteredStream = SettingsManager.getInstance().getSearchFuzzy() ? applyJaroWinklerAlbumFilter(albumStream) : applyAlbumFilter(albumStream);
                    observer.onSuccess(filteredStream.toList());
                }

                @Override
                public void onError(Throwable e) {
                    observer.onError(e);
                }
            };
        }

        Stream<Album> applyJaroWinklerAlbumFilter(Stream<Album> stream) {
            return stream.map(album -> new SearchUtils.JaroWinklerObject<>(album, filterString, album.name))
                    .filter(jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString))
                    .sorted((a, b) -> a.object.compareTo(b.object))
                    .sorted((a, b) -> Double.compare(b.score, a.score))
                    .map(jaroWinklerObject -> jaroWinklerObject.object);
        }

        Stream<Album> applyAlbumFilter(Stream<Album> stream) {
            return stream.filter(album -> StringUtils.containsIgnoreCase(album.name, filterString));
        }
    }

    private class AlbumArtistFilterOperator implements SingleOperator<List<AlbumArtist>, List<AlbumArtist>> {

        String filterString;

        AlbumArtistFilterOperator(@NonNull String filterString) {
            this.filterString = filterString;
        }

        @Override
        public SingleObserver<? super List<AlbumArtist>> apply(SingleObserver<? super List<AlbumArtist>> observer) {
            return new SingleObserver<List<AlbumArtist>>() {
                @Override
                public void onSubscribe(Disposable d) {
                    observer.onSubscribe(d);
                }

                @Override
                public void onSuccess(List<AlbumArtist> albumArtists) {
                    Collections.sort(albumArtists, AlbumArtist::compareTo);
                    Stream<AlbumArtist> albumArtistStream = Stream.of(albumArtists).filter(albumArtist -> albumArtist.name != null);
                    Stream<AlbumArtist> filteredStream =
                            SettingsManager.getInstance().getSearchFuzzy() ? applyJaroWinklerAlbumArtistFilter(albumArtistStream) : applyAlbumArtistFilter(albumArtistStream);
                    observer.onSuccess(filteredStream.toList());
                }

                @Override
                public void onError(Throwable e) {
                    observer.onError(e);
                }
            };
        }

        Stream<AlbumArtist> applyJaroWinklerAlbumArtistFilter(Stream<AlbumArtist> stream) {
            return stream.map(albumArtist -> new SearchUtils.JaroWinklerObject<>(albumArtist, filterString, albumArtist.name))
                    .filter(jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString))
                    .sorted((a, b) -> a.object.compareTo(b.object))
                    .sorted((a, b) -> Double.compare(b.score, a.score))
                    .map(jaroWinklerObject -> jaroWinklerObject.object);
        }

        Stream<AlbumArtist> applyAlbumArtistFilter(Stream<AlbumArtist> stream) {
            return stream.filter(albumArtist -> StringUtils.containsIgnoreCase(albumArtist.name, filterString));
        }
    }
}

File: app/src/main/java/com/simplecity/amp_library/ui/screens/search/SearchPresenter.kt
package com.simplecity.amp_library.ui.screens.search

import android.text.TextUtils
import com.simplecity.amp_library.data.Repository
import com.simplecity.amp_library.model.Album
import com.simplecity.amp_library.model.AlbumArtist
import com.simplecity.amp_library.model.Song
import com.simplecity.amp_library.playback.MediaManager
import com.simplecity.amp_library.ui.common.Presenter
import com.simplecity.amp_library.ui.modelviews.AlbumArtistView
import com.simplecity.amp_library.ui.modelviews.AlbumView
import com.simplecity.amp_library.ui.screens.album.menu.AlbumArtistMenuContract
import com.simplecity.amp_library.ui.screens.album.menu.AlbumArtistMenuPresenter
import com.simplecity.amp_library.ui.screens.album.menu.AlbumMenuContract
import com.simplecity.amp_library.ui.screens.album.menu.AlbumMenuPresenter
import com.simplecity.amp_library.ui.screens.songs.menu.SongMenuContract
import com.simplecity.amp_library.ui.screens.songs.menu.SongMenuPresenter
import com.simplecity.amp_library.utils.LogUtils
import com.simplecity.amp_library.utils.SettingsManager
import com.simplecity.amp_library.utils.StringUtils
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.SingleOperator
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function3
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

class SearchPresenter @Inject
constructor(
        private val mediaManager: MediaManager,
        private val songsRepository: Repository.SongsRepository,
        private val albumsRepository: Repository.AlbumsRepository,
        private val albumArtistsRepository: Repository.AlbumArtistsRepository,
        private val settingsManager: SettingsManager,
        private val songMenuPresenter: SongMenuPresenter,
        private val albumMenuPresenter: AlbumMenuPresenter,
        private val albumArtistsMenuPresenter: AlbumArtistMenuPresenter

) : Presenter<SearchView>(),
        SongMenuContract.Presenter by songMenuPresenter,
        AlbumMenuContract.Presenter by albumMenuPresenter,
        AlbumArtistMenuContract.Presenter by albumArtistsMenuPresenter {

    private var performSearchSubscription: Disposable? = null

    private var query: String? = null

    override fun bindView(view: SearchView) {
        super.bindView(view)

        songMenuPresenter.bindView(view)
        albumMenuPresenter.bindView(view)
        albumArtistsMenuPresenter.bindView(view)

        view.setFilterFuzzyChecked(settingsManager.searchFuzzy)
        view.setFilterArtistsChecked(settingsManager.searchArtists)
        view.setFilterAlbumsChecked(settingsManager.searchAlbums)
    }

    override fun unbindView(view: SearchView) {
        super.unbindView(view)
        songMenuPresenter.unbindView(view)
        albumMenuPresenter.unbindView(view)
        albumArtistsMenuPresenter.unbindView(view)
    }

    fun queryChanged(query: String?) {
        var query = query

        if (TextUtils.isEmpty(query)) {
            query = ""
        }

        if (query == this.query) {
            return
        }

        loadData(query!!)

        this.query = query
    }

    private fun loadData(query: String) {

        val searchView = view

        if (searchView != null) {

            searchView.setLoading(true)

            //We've received a new refresh call. Unsubscribe the in-flight subscription if it exists.
            if (performSearchSubscription != null) {
                performSearchSubscription!!.dispose()
            }

            val albumArtistsObservable = if (settingsManager.searchArtists)
                albumArtistsRepository.getAlbumArtists()
                        .first(emptyList())
                        .lift(AlbumArtistFilterOperator(query))
            else
                Single.just(emptyList())

            val albumsObservable = if (settingsManager.searchAlbums)
                albumsRepository.getAlbums()
                        .first(emptyList())
                        .lift(AlbumFilterOperator(query))
            else
                Single.just(emptyList())

            val songsObservable = songsRepository.getSongs(null as Function1<Song, Boolean>?)
                    .first(emptyList())
                    .lift(SongFilterOperator(query))

            performSearchSubscription = Single.zip<List<AlbumArtist>, List<Album>, List<Song>, SearchResult>(albumArtistsObservable, albumsObservable, songsObservable, Function3 { albumArtists: List<AlbumArtist>, albums: List<Album>, songs: List<Song> -> SearchResult(albumArtists, albums, songs) })
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            { searchView.setData(it) },
                            { error -> LogUtils.logException(TAG, "Error refreshing adapter", error) }
                    )

            addDisposable(performSearchSubscription!!)
        }
    }

    fun setSearchFuzzy(searchFuzzy: Boolean) {
        settingsManager.searchFuzzy = searchFuzzy
        loadData(query!!)
    }

    fun setSearchArtists(searchArtists: Boolean) {
        settingsManager.searchArtists = searchArtists
        loadData(query!!)
    }

    fun setSearchAlbums(searchAlbums: Boolean) {
        settingsManager.searchAlbums = searchAlbums
        loadData(query!!)
    }

    fun onSongClick(songs: List<Song>, song: Song) {
        val view = view

        mediaManager.playAll(songs, songs.indexOf(song), true) {
            view?.showPlaybackError()
            Unit
        }
    }

    fun onArtistClicked(albumArtistView: AlbumArtistView, viewholder: AlbumArtistView.ViewHolder) {
        val view = view
        view?.goToArtist(albumArtistView.albumArtist, viewholder.imageOne)
    }

    fun onAlbumClick(albumView: AlbumView, viewHolder: AlbumView.ViewHolder) {
        val view = view
        view?.goToAlbum(albumView.album, viewHolder.imageOne)
    }

    private inner class SongFilterOperator internal constructor(internal var filterString: String) : SingleOperator<List<Song>, List<Song>> {

        override fun apply(observer: SingleObserver<in List<Song>>): SingleObserver<in List<Song>> {
            return object : SingleObserver<List<Song>> {
                override fun onSubscribe(d: Disposable) {
                    observer.onSubscribe(d)
                }

                override fun onSuccess(songs: List<Song>) {
                    var songs = songs
                    val songList = songs.filter { song -> song.name != null }
                    songs = (if (settingsManager.searchFuzzy) applyJaroWinklerFilter(songList) else applySongFilter(songList)).toList()
                    observer.onSuccess(songs)
                }

                override fun onError(e: Throwable) {
                    observer.onError(e)
                }
            }
        }

        internal fun applyJaroWinklerFilter(songList: List<Song>): List<Song> {
            return songList.map { song -> JaroWinklerObject(song, filterString, song.name) }
                    .filter { jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString) }
                    .sortedWith(Comparator { a, b -> a.`object`.compareTo(b.`object`) })
                    .sortedWith(Comparator { a, b -> java.lang.Double.compare(b.score, a.score) })
                    .map { jaroWinklerObject -> jaroWinklerObject.`object` }
        }

        internal fun applySongFilter(songStream: List<Song>): List<Song> {
            return songStream.filter { song -> StringUtils.containsIgnoreCase(song.name, filterString) }
        }
    }

    private inner class AlbumFilterOperator internal constructor(internal var filterString: String) : SingleOperator<List<Album>, List<Album>> {

        override fun apply(observer: SingleObserver<in List<Album>>): SingleObserver<in List<Album>> {
            return object : SingleObserver<List<Album>> {
                override fun onSubscribe(d: Disposable) {
                    observer.onSubscribe(d)
                }

                override fun onSuccess(albums: List<Album>) {
                    albums.sortedWith(Comparator { a, b -> a.compareTo(b) })
                    val albumStream = albums.filter { album -> album.name != null }
                    val filteredStream = if (settingsManager.searchFuzzy) applyJaroWinklerAlbumFilter(albumStream) else applyAlbumFilter(albumStream)
                    observer.onSuccess(filteredStream.toList())
                }

                override fun onError(e: Throwable) {
                    observer.onError(e)
                }
            }
        }

        internal fun applyJaroWinklerAlbumFilter(albums: List<Album>): List<Album> {
            return albums.map { album -> JaroWinklerObject(album, filterString, album.name) }
                    .filter { jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString) }
                    .sortedWith(Comparator { a, b -> a.`object`.compareTo(b.`object`) })
                    .sortedWith(Comparator { a, b -> java.lang.Double.compare(b.score, a.score) })
                    .map { jaroWinklerObject -> jaroWinklerObject.`object` }
        }

        internal fun applyAlbumFilter(stream: List<Album>): List<Album> {
            return stream.filter { album -> StringUtils.containsIgnoreCase(album.name, filterString) }
        }
    }

    private inner class AlbumArtistFilterOperator internal constructor(internal var filterString: String) : SingleOperator<List<AlbumArtist>, List<AlbumArtist>> {

        override fun apply(observer: SingleObserver<in List<AlbumArtist>>): SingleObserver<in List<AlbumArtist>> {
            return object : SingleObserver<List<AlbumArtist>> {
                override fun onSubscribe(d: Disposable) {
                    observer.onSubscribe(d)
                }

                override fun onSuccess(albumArtists: List<AlbumArtist>) {
                    Collections.sort(albumArtists) { obj, albumArtist -> obj.compareTo(albumArtist) }
                    val albumArtistList = albumArtists.filter { albumArtist -> albumArtist.name != null }
                    val filteredList = if (settingsManager.searchFuzzy) applyJaroWinklerAlbumArtistFilter(albumArtistList) else applyAlbumArtistFilter(albumArtistList)
                    observer.onSuccess(filteredList.toList())
                }

                override fun onError(e: Throwable) {
                    observer.onError(e)
                }
            }
        }

        internal fun applyJaroWinklerAlbumArtistFilter(stream: List<AlbumArtist>): List<AlbumArtist> {
            return stream.map { albumArtist -> JaroWinklerObject(albumArtist, filterString, albumArtist.name) }
                    .filter { jaroWinklerObject -> jaroWinklerObject.score > SCORE_THRESHOLD || TextUtils.isEmpty(filterString) }
                    .sortedWith(Comparator { a, b -> a.`object`.compareTo(b.`object`) })
                    .sortedWith(Comparator { a, b -> java.lang.Double.compare(b.score, a.score) })
                    .map { jaroWinklerObject -> jaroWinklerObject.`object` }
        }

        internal fun applyAlbumArtistFilter(stream: List<AlbumArtist>): List<AlbumArtist> {
            return stream.filter { albumArtist -> StringUtils.containsIgnoreCase(albumArtist.name, filterString) }
        }
    }

    override fun <T> transform(src: Single<List<T>>, dst: (List<T>) -> Unit) {
        addDisposable(
                src
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe(
                                { items -> dst(items) },
                                { error -> LogUtils.logException(SearchPresenter.TAG, "Failed to transform src single", error) }
                        )
        )
    }

    companion object {

        private const val TAG = "SearchPresenter"

        private const val SCORE_THRESHOLD = 0.80
    }
}
